src/gpu/multiexp.rs

```diff
@@ -287,44 +292,64 @@ where
+            if n > 0 {
                 bases
-                    .par_chunks(chunk_size)
-                    .zip(exps.par_chunks(chunk_size))
-                    .zip(self.kernels.par_iter_mut())
-                    .map(|((bases, exps), kern)| -> Result<<G as CurveAffine>::Projective, GPUError> {
-                        let mut acc = <G as CurveAffine>::Projective::zero();
-                        for (bases, exps) in bases.chunks(kern.n).zip(exps.chunks(kern.n)) {
-                            let result = kern.multiexp(bases, exps, bases.len())?;
-                            acc.add_assign(&result);
-                        }
-
-                        Ok(acc)
-                    })
-                    .collect::<Vec<_>>()
+                .par_chunks(chunk_size)
+                .zip(exps.par_chunks(chunk_size))
+                .zip(self.kernels.par_iter_mut())
+                .map(|((bases, exps), kern)| -> Result<<G as CurveAffine>::Projective, GPUError> {
+                    let mut acc = <G as CurveAffine>::Projective::zero();
+                    for (bases, exps) in bases.chunks(kern.n).zip(exps.chunks(kern.n)) {
+                        let result = kern.multiexp(bases, exps, bases.len())?;
+                        acc.add_assign(&result);
+                    }
+
+                    Ok(acc)
+                })
+                .collect::<Vec<_>>()
             } else {
                 Vec::new()
             };
 
-            let cpu_acc = cpu_multiexp(
-                &pool,
-                (Arc::new(cpu_bases.to_vec()), 0),
-                FullDensity,
-                Arc::new(cpu_exps.to_vec()),
-                &mut None,
-            );
+        let cpu_acc = cpu_multiexp(
+            &pool,
+            (Arc::new(cpu_bases.to_vec()), 0),
+            FullDensity,
+            Arc::new(cpu_exps.to_vec()),
+            &mut None,
+        );
 
-            for r in results {
-                acc.add_assign(&r?);
-            }
+        for r in results {
+            acc.add_assign(&r.unwrap());
+        }
 
-            acc.add_assign(&cpu_acc.wait().unwrap());
-            Ok(acc)
-        })
+        (Ok(acc), cpu_acc)
+    }
+
+    pub fn multiexp<G>(
+        &mut self,
+        pool: &Worker,
+        bases: Arc<Vec<G>>,
+        exps: Arc<Vec<<<G::Engine as ScalarEngine>::Fr as PrimeField>::Repr>>,
+        skip: usize,
+        n: usize,
+    ) -> GPUResult<<G as CurveAffine>::Projective>
+    where
+        G: CurveAffine,
+        <G as groupy::CurveAffine>::Engine: crate::bls::Engine,
+    {
+        // NOTE: This is called in the context of the THREAD_POOL
+        // already.  It's not clear there's any benefit (or harm) to
+        // doing this again.
+        let (acc, cpu_acc) = crate::multicore::THREAD_POOL
+            .install(|| self.multiexp_inner(pool, bases, exps, skip, n));
+
+        let mut unwrapped_acc = acc?;
+        unwrapped_acc.add_assign(&cpu_acc.wait().unwrap());
```

Comment:
Since this method is called in the context of the THREAD_POOL already, this `cpu_acc.wait()` can still potentially break with a race.

https://github.com/filecoin-project/bellperson/blob/d2e88544efb9876bc96c3d7e26a527943c0fdb68/src/multiexp.rs#L313